src/Microsoft.AspNetCore.Server.IIS/IISHttpRequestBody.cs

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Server.IIS
+{
+    internal class IISHttpRequestBody : Stream
+    {
+        private readonly HttpProtocol _httpContext;
+
+        public IISHttpRequestBody(HttpProtocol httpContext)
+        {
+            _httpContext = httpContext;
+        }
+
+        public override bool CanRead => true;
+
+        public override bool CanSeek => false;
+
+        public override bool CanWrite => false;
+
+        public override long Length => throw new NotSupportedException();
+
+        public override long Position { get => throw new NotSupportedException(); set => throw new NotSupportedException(); }
+
+        public override void Flush()
+        {
+            throw new NotSupportedException();
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            return ReadAsync(buffer, offset, count, CancellationToken.None).GetAwaiter().GetResult();
+        }
+
+        public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            _httpContext.StartReadingRequestBody();
+
+            while (true)
+            {
+                var result = await _httpContext.Input.Reader.ReadAsync();
+                var readableBuffer = result.Buffer;
+                try
+                {
+                    if (!readableBuffer.IsEmpty)
+                    {
+                        var actual = Math.Min(readableBuffer.Length, count);
+                        readableBuffer = readableBuffer.Slice(0, actual);
+                        readableBuffer.CopyTo(buffer);
+                        return (int)actual;
+                    }
+                    else if (result.IsCompleted)
+                    {
+                        return 0;
+                    }
+                }
+                finally
+                {
+                    _httpContext.Input.Reader.Advance(readableBuffer.End, readableBuffer.End);
```

Comment:
What about 'actual'?

Reply:
Yeah I believe that should be advancing by 'actual' and not to the end of the buffer, right?